### PR TITLE
Hubble: add no-metrics label suport for namespace or pod 

### LINKED ISSUE #30788: CFP: Hubble Label to exclude namespaces/pods from metrics
## Cilium Feature Proposal

**Is your proposed feature related to a problem?**

Managing Hubble metrics cardinality helps reduce memory being used within the pod, while ingesting and also to store. #23162  did introduce some improvements, this CFP is an extension to one of the points laid out in that issue. 

In cloud providers, metrics have costs at multiple levels, while ingesting, storing and also querying. Providing options to actively reduce the number of active time series users can persist will help in memory and cost savings. 

**Describe the feature you'd like**

This proposal is re-hashing what was described as bullet 4 in #23162. introducing a new label to exclude namespaces/pods from metrics.

`hubble.cilium.io/no-metrics: "true"` 

Applying this label on:

1) Namespace: Since this label gets applied on each endpoint inside this namespace, we can ignore all the endpoints to be included in hubble context label generation
2) Pods: we can ignore context label generation for this endpoint alone.

### PR DESCRIPTION
**Pull request by vakalapa**
This change will introduce a new hubble label which can be applied on namespace or pod. Hubble while generating context labels for a given metrics, will check for the given src/dst endpoint inside the flow for this label, if the label is present, then hubble will ignore generating labels for this endpoints and will return empty label.

``hubble.cilium.io/no-metrics: "true"``

Applying this label on:

- Namespace: Since this label gets applied on each endpoint inside this namespace, we can ignore all the endpoints to be included in hubble context label generation
- Pods: we can ignore context label generation for this endpoint alone.

Fixes: #30788 , #23162

### CODE REVIEW

File: pkg/hubble/metrics/api/context.go [hunk 1]
```diff
@@ -304,35 +309,55 @@ func labelsContext(invertSourceDestination bool, wantedLabels labelsSet, flow *p
 			case "source_ip":
```
Comment (vakalapa) [inline pkg/hubble/metrics/api/context.go:309]: i was thinking sometimes customers would want to see the IP if endpoints are explicitly ignored, but on second thought, hubble should ignore adding IPs too?

File: Documentation/observability/metrics.rst [hunk 2]
```diff
@@ -155,6 +155,20 @@ have it scrape all Hubble metrics from the endpoints automatically:
 
 .. _hubble_open_metrics:
 
+Excluding Namespaces/Pods from Hubble Metrics
+--------------------------------------------
```
Comment (vakalapa) [inline Documentation/observability/metrics.rst:159]: fix: observability/metrics.rst:159: (WARNING/2) Title underline too short.

**Comment by chancez** [comment]
I'm not sure this is the best approach for this. Instead, normally you should be handling this in your metrics collector. In prometheus you can drop metrics based on labels which would allow for you to do this, and a lot more, with more flexibility. I'm not sure that the added complexity here is warranted when there's a reasonable solution to doing this outside of Hubble.

**Comment by vakalapa** [comment]
@chancez  true normally in unmanaged Prometheus scenarios, metrics can be dropped at ingestion. But in managed cloud Prometheus scenarios without dynamic metrics configuration support, it becomes tedious for customers to manage this? Also at scale with large number of active time series, we can reduce load on prometheus collector with this approach?

**Comment by chancez** [comment]
> @chancez true normally in unmanaged Prometheus scenarios, metrics can be dropped at ingestion. But in managed cloud Prometheus scenarios without dynamic metrics configuration support, it becomes tedious for customers to manage this? 

I'm not entirely sure what you mean here by dynamic metrics, but in all of the major managed Prometheus products I'm aware of (AWS, GCP, Grafana), they all support relabeling which means you can drop metrics at scrape time. 

> Also at scale with large number of active time series, we can reduce load on prometheus collector with this approach?

That's a valid consideration, dropping at scrape time, does still mean scraping it, and that is additional processing the collector has to do. 

So given that, I can see some validity to the idea. However, I think the implementation still has some problems:

As mentioned, this only changes the labels. This doesn't actually change whether or not we process metrics for a given pod. This is pretty confusing IMO. The original proposal in https://github.com/cilium/cilium/issues/23162 was to exclude pods/namespaces from metrics entirely, not simply to adjust the labels on the metrics.

Simply adjusting the labels means pods/namespaces still contribute to a metric in aggregate. This is going to result in confusing and hard to use metrics because there will be missing dimensions in some cases. It may even break metrics in some cases.

I would consider a feature that actually disables metrics accounting for specific pods or namespaces. If that's something you would be interested in implementing then here's a few considerations:

1. Configuring what pods contribute to metrics feels similar to other features requests to only export flows for specific pods/namespaces. We currently support filtering flows on export, and do not support filtering on metrics at all. I think it would be ideal to have a similar approach to metrics. Perhaps we should consider a config file similar to what was added for [dynamic flow export](https://github.com/cilium/cilium/pull/28873). 
2. Configuring metrics via pod/namespace labels sounds nice, but in practice, it makes it difficult to manage things centrally. Admins may not want users to be able to define if Hubble metrics are collected for their pod. I think having a config file makes more sense, allowing the administrators to decide what's collected. If you wanted to still make it possible to configure it by labels, then I think the config file should define which labels to consider (see below).
3. I think we should be using FlowFilter to filter out flows before they get to the metrics plugins. This is a common method of configuring Flow related functionality, and already has a ton of flexibility, you should even be able to use Flow filters in the proposed config file to only expose metrics for pods with a given label, as you desire.

**Comment by vakalapa** [comment]
>  this only changes the labels. This doesn't actually change whether or not we process metrics for a given pod. 

I was trying to reduce number of active time series being generated by Hubble and to preserve the total counts of flows/bytes/flags etc. that was if some users are relying on the total/sum counts of events/flags those could still be preserved. In a way, we are accounting these ignored metrics at a node level aggregation ? 

> This is going to result in confusing and hard to use metrics because there will be missing dimensions in some cases. 

User would/should know what pods/namespaces they have excluded, do you think they might unintentionally label and it could cause to problems ?

> It may even break metrics in some cases.

Can you help me understand when/what cases it could be, I might be naive in thinking this would be straightforward.

**Comment by chancez** [comment]
> > this only changes the labels. This doesn't actually change whether or not we process metrics for a given pod.
> 
> I was trying to reduce number of active time series being generated by Hubble and to preserve the total counts of flows/bytes/flags etc. that was if some users are relying on the total/sum counts of events/flags those could still be preserved. In a way, we are accounting these ignored metrics at a node level aggregation ?
> 
> > This is going to result in confusing and hard to use metrics because there will be missing dimensions in some cases.
> 
> User would/should know what pods/namespaces they have excluded, do you think they might unintentionally label and it could cause to problems ?

In general, this is not generally not how Prometheus is used. Having some targets which have some labels but not others is just not common practice, and not generally the right way to go about this. IMO if this cardinality is the concern, using relabeling and recording rules is the correct solution.

> > It may even break metrics in some cases.
> 
> Can you help me understand when/what cases it could be, I might be naive in thinking this would be straightforward.

I just took another look, it shouldn't _break_ anything, since you're using an empty string as the value, I was thinking about the case where not all labels are supplied, as that would cause a panic in the prometheus instrumentation.

**Comment by vakalapa** [comment]
@chancez apologies for the delayed response,

For dynamic flow filter based approach, do you think having a single config map which allows only one filter config at a given time, which can then be used to filter out flows in below line would be a better approach?

https://github.com/cilium/cilium/blob/5864db744bb2a85e6ef9d026262c77dae8bf9f50/daemon/cmd/hubble.go#L148-L154

This could be a breaking change, so i am assuming it should be wrapped around a configmap option. If the configmap options say `hubble-metrics-dynamic-filter` is turned on, then by default metrics should not be accounted for unless a flow matches an allow/deny list in new config map called ?

if this is a valid approach, one question i have is how to manage previous time series which no longer satisfy a newer dynamic flow filter, is it possible to soft restart hubble metrics to start from scratch again ?

**Comment by chancez** [comment]
## New config option

> @chancez apologies for the delayed response,
> 
> For dynamic flow filter based approach, do you think having a single config map which allows only one filter config at a given time, which can then be used to filter out flows in below line would be a better approach?
>
> ...
> 
> if this is a valid approach, one question i have is how to manage previous time series which no longer satisfy a newer dynamic flow filter, is it possible to soft restart hubble metrics to start from scratch again ?

Two things, generally: I think that's a decent idea as well, and after a re-review: I'm realizing this statement I made:

> As mentioned in your PR description, this only changes the labels. This doesn't actually change whether or not we process metrics for a given pod. I believe this is pretty confusing and unexpected behavior though. The original proposal in https://github.com/cilium/cilium/issues/23162 was to exclude pods/namespaces from metrics entirely, not simply to adjust the labels on the metrics.

Wasn't a correct statement. I see now that it's doing what was originally described but implementation confused me in `labelsContext` thinking it was only changing the metrics labels. I can do a proper PR review, however, I think it might be a good idea to write a design proposal before going too deep into a specific implementation (see more about this towards the end of my comment).

Going back to your idea of adding a new config option:

This seems generally useful and I would be open to the idea of that too. We have some logic to handle deleting series based on the pod already from https://github.com/cilium/cilium/pull/23385, but if the configuration is set statically (eg via a flag), then it wouldn't matter and restarting the cilium agent pod would be needed to reapply configuration, which will also reset all series.

However, one thing that feels useful would be the ability to specify the allow/deny list on a per-enabled-metric basis. This is currently done using [contextOptions](https://docs.cilium.io/en/stable/observability/metrics/#context-options) you can specify in the metrics list. That being said, I have mixed feelings about adding more to `contextOptions`, especially something fairly structured like `FlowFilters`, because we're putting a lot of structured configuration inside of a simple string, and this isn't scaling very well as we add new metrics features. I'm not even sure it will work with the current naive parser, so we may want something better, like what was done in #28873 (see the next section).

## Similar features

With a feature like what's being suggested, the metrics feature is starting to look similar to how the static and dynamic flow file export works, which lets you filter which flows you want to export via an allow/deny list of FlowFilters.

A contributor recently added support to configure flow export using a [config file](https://github.com/cilium/cilium/pull/28873) which is reloaded periodically to allow runtime reconfiguration of export. I wonder if we should consider doing something similar for metrics configuration, especially since the flag based approaches for these complex configurations are starting to get out of hand.

## TLDR

All of that is leading back to: Metrics configurations are starting to get pretty complicated as we add more features, so we may want to take a step back and write a proper CFP for how we want to handle advanced metrics configurations like the one you've proposed. This would include the different approaches and their trade-offs. Eg: adding a new flag vs adding a new contextOption vs doing it via pod labels vs adding a new metrics config file, etc. I also know at one point the Hubble team discussed doing this via CNPs like we do for other options. 

Would you be willing to write a design proposal covering some of these different approaches and the pros/cons to each? Then we can involve more of the team and review it before deciding on a specific implementation.

**Comment by vakalapa** [comment]
Sounds good, fwiw, I was modelling my idea of dynamic configmap for metrics config based on #23385. I can take a stab at putting together a CFP doc. I am asumming https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#cilium-feature-proposals is the right process to follow ?

**Comment by chancez** [comment]
The process for CFPs has changed a little bit, we should be creating PRs to https://github.com/cilium/design-cfps/ and using https://github.com/cilium/design-cfps/blob/main/cfps/CFP-003-template.md as a template.

**Comment by chancez** [comment]
@vakalapa Do you mind if we close this while the CFP is in progress? It can always be re-opened if we choose to revisit this approach. I figure we don't need to distract reviewers with the PR until we do the CFP.

**Comment by vakalapa** [comment]
For sure thanks for the reminder, i will be raising the CFP in a couple of days and link it to this PR.

**Comment by vakalapa** [comment]
@chancez  created the below CFP, please let me know how this one looks ?

https://github.com/cilium/design-cfps/pull/29
